File: [0091][Medium][Decode_Ways]/Decode_Ways.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 36 ms
# leetcode memory   cost : 13.7 MB
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        # dp[i] is when come to i element s[i-1], the possible decode kinds
        dp = [0]*(n+1)

        # no need check none since s is a non-empty string
        if n>=1: 
            dp[0] = 1 

        # start with more than 3 bit, loop i from 2 to (n)
        for i in range(1,n+1):

            # important,check whether s[i] can be a individual char, if yes, need add possible of dp[i-1]
            # if xx0xx, the 0 must be combined with before , should not add dp[i-1]       
            if 1<= int(s[i-1]) <= 9:
                dp[i] += dp[i-1] 

            # check whether s[i] can be a combined char, if yes, need add possible of  dp[i-2]       
            # wrong example: if  i>1 and 1<= int(s[i-2]) <= 2 and 0<= int(s[i-1]) <= 6:
            if  i>=2 and 10 <= int(s[i-2])*10 + int(s[i-1]) <= 26:
                dp[i] +=  dp[i-2]

        print(dp)                
        return dp[n]

def main():
    code = "226"             #expect is 3
    obj = Solution()
    result = obj.numDecodings(code)
    print("return result is :",result)
    
if __name__ =='__main__':
    main()   

File: [0091][Medium][Decode_Ways]/Decode_Ways.java
// leetcode time     cost : 7 ms
// leetcode memory   cost : 43.5 MB
class Solution {
    public int numDecodings(String s) {
            if (s == null || s.length() == 0) {
                return 0;
            }
            return DFS(s, 0);
        }
    
        private int DFS(String s, int start) {
            //递归的第一步，应该是加终止条件，避免死循环。
            if (s.length() == start) {
                return 1;
            }
            //以0位开始的数是不存在的
            if (s.charAt(start) == '0') {
                return 0;
            }
            //递归的递推式应该是如果index的后两位小于等于26，  
            int ans1 = DFS(s, start + 1);
            int ans2 = 0;
            if (start < s.length() - 1) {
                int ten = (s.charAt(start) - '0') * 10;
                int one = (s.charAt(start + 1) - '0');
                if (ten + one <= 26) {
                    ans2 = DFS(s, start + 2);
                }
            }
            return ans1 + ans2;
        }
    }

public class Decode_Ways {
    public static void main(String args[]) {
        String word = "172426543102320";                     // #expect is 16
        Solution Solution_obj = new Solution();
        int result = Solution_obj.numDecodings(word);
        System.out.println("return value is :" + (result));
    }
}